src/runner/browser-env/vite/browser-modules/driver.ts

```diff
@@ -0,0 +1,173 @@
+    SeleniumProtocol,
+    GeckoProtocol,
+    WebDriverBidiProtocol,
+} from "@wdio/protocols";
+import { webdriverMonad, sessionEnvironmentDetector, devtoolsEnvironmentDetector } from "@wdio/utils";
+import { getEnvironmentVars } from "webdriver";
+import { SOCKET_MAX_TIMEOUT, SOCKET_TIMED_OUT_ERROR, MAX_ARGS_LENGTH } from "./constants.js";
+import { BrowserEventNames } from "./types.js";
+
+type ProtocolCommandFn = (...args: unknown[]) => Promise<unknown>;
+type PropertiesObject = Record<string, PropertyDescriptor>;
+
+const COMMANDS_HANDLED_BY_SERVER = ["debug", "saveScreenshot", "savePDF"];
+
+export default class ProxyDriver {
+    static newSession(
+        params: Record<string, unknown>,
+        modifier: never,
+        userPrototype: Record<string, PropertyDescriptor>,
+        commandWrapper: VoidFunction | undefined,
+    ): unknown {
+        const monad = webdriverMonad(params, modifier, getWdioPrototype(userPrototype));
+        return monad(window.__testplane__.sessionId, commandWrapper);
+    }
+}
+
+function getWdioPrototype(userPrototype: Record<string, PropertyDescriptor>): PropertiesObject {
+    for (const commandName of COMMANDS_HANDLED_BY_SERVER) {
+        delete userPrototype[commandName];
+    }
+
+    const prototype = {
+        ...getMockedProtocolCommands(),
+        ...getEnvironmentFlags(),
+        ...userPrototype,
+    };
+
+    return prototype;
+}
+
+function getAllProtocolCommands(): string[] {
+    const protocols = [
+        WebDriverProtocol,
+        MJsonWProtocol,
+        JsonWProtocol,
+        AppiumProtocol,
+        ChromiumProtocol,
+        SauceLabsProtocol,
+        SeleniumProtocol,
+        GeckoProtocol,
+        WebDriverBidiProtocol,
+    ];
+    const commands = new Set<string>();
+
+    for (const protocol of protocols) {
+        for (const endpointData of Object.values(protocol)) {
+            for (const { command } of Object.values<{ command: string }>(endpointData)) {
+                commands.add(command);
+            }
+        }
+    }
+
+    return Array.from(commands);
+}
+
+function getMockedProtocolCommands(): PropertiesObject {
+    return [...getAllProtocolCommands(), ...COMMANDS_HANDLED_BY_SERVER, ...window.__testplane__.customCommands].reduce(
+        (prev, commandName) => {
+            prev[commandName] = { value: mockCommand(commandName) };
+            return prev;
+        },
+        {} as PropertiesObject,
+    );
+}
+
+function mockCommand(commandName: string): ProtocolCommandFn {
+    return async (...args: unknown[]): Promise<unknown> => {
+        const { socket } = window.__testplane__;
+        const timeout = getCommandTimeout(commandName);
+
+        try {
+            // TODO: remove type casting after https://github.com/socketio/socket.io/issues/4925
+            const [error, result] = (await socket
+                .timeout(timeout)
+                .emitWithAck(BrowserEventNames.runBrowserCommand, { name: commandName, args })) as [
```

Comment:
Send command to the server (testplane master process) which passes it to worker process and execute command.